### DISCUSSION TITLE
v13 to v15 on cloud but gives erro ksa vat module not found

### DISCUSSION THREAD

**Discussion by ashish-greycube** [lavaloon-eg/ksa_compliance#162] [Q&A]
My site is v13 saudi arabi and migrated to v15
on v15 my sales invoice print formats i.e.
Flex VAT
Flex VAT QR Up
SI QR up No Disc
KSA VAT Format
Flex VAT QR Down
Simplified VAT Invoice
etc donot work, it gives erro ksa vat module not found

in v15 i understand that ksa specific code is shifted out of erpnext
first we had https://github.com/frappe/ksa
now https://github.com/lavaloon-eg/ksa_compliance

Q1 : which app to use ?
Q2 : i tried to use lavaloon-eg/ksa_compliance but the earlier print formats i.e.
Flex VAT
Flex VAT QR Up
SI QR up No Disc
KSA VAT Format
Flex VAT QR Down
Simplified VAT Invoicet
donot work. Also its print format i.e. zatca_phase_1_print_format also give error
![image](https://github.com/user-attachments/assets/17575a2c-0c60-4da8-8a81-c9a29bfa4d14)
![image](https://github.com/user-attachments/assets/504d9258-09f5-48db-9bc2-bd76df7b6a54)

**Comment by mhaggag** [comment]
Hello,

If these were added by the old ERPNext KSA-specific logic that was extracted to its own in v14, then it'd be in [frappe's KSA app](https://github.com/frappe/ksa), which only supports phase 1.

Our app, ksa_compliance, is a separate app with its own print formats, and it supports phase 1 and 2. If you'd like to try it out, we recommend using a test/staging instance first to run on the ZATCA sandbox, then simulation, before going into production. There is some documentation on the [app's wiki](https://lavaloon.com/wiki/E-INVOICE%20KSA%20(Zatca%20))

**Reply by ashish-greycube** [reply]
@mhaggag thanks for quick reply.
Can the two apps frappe/ksa and your app coexist ? thanks

**Reply by mhaggag** [reply]
We do not support it in the sense that we don't test for it or fix issues that may come up, but as far as I can see the two apps may co-exist. They do try to create the same custom fields on sales invoice item and POS invoice item, but I think frappe will take the field definitions of one app over the other and it'll work.

If you do try it, please make sure you do it on a testing/staging clone of your production instances to avoid trouble.

**Comment by ashish-greycube** [comment]
@mhaggag , thanks for reply
does your app generate the QR code that used happen with ksa app earlier
I mean this part
![image](https://github.com/user-attachments/assets/a2216ba4-a1d3-411a-9113-d319321003de)
![image](https://github.com/user-attachments/assets/20c8a327-bd8c-40dc-b8c1-93e3a7267e09)

**Reply by mhaggag** [reply]
We do, but we generate it on demand instead of attaching it to the invoice to avoid cluttering system files and attachments.

For phase 1, the print formats we use "ZATCA Phase 1 Print Format" and "ZATCA Phase 1 Print Format - POS Invoice" call a jinja function to generate the QR code.

For phase 2, it's a virtual field on `Sales Invoice Additional Fields` (which represents the ZATCA mapping for a given invoice) that is also generated in memory on-demand.

**Reply by ashish-greycube** [reply]
"ZATCA Phase 1 Print Format"  gives error
![image](https://github.com/user-attachments/assets/c4eabbd4-cc9b-40f2-aa58-e8fd61a443b5)

  File "apps/frappe/frappe/utils/pdf.py", line 49, in pdf_body_html
    return template.render(args, filters={"len": len})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "env/lib/python3.11/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "env/lib/python3.11/site-packages/jinja2/environment.py", line 939, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
TypeError: cannot unpack non-iterable NoneType object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 114, in application
    response = frappe.api.handle(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/api/__init__.py", line 49, in handle
    data = endpoint(**arguments)
           ^^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/api/v1.py", line 36, in handle_rpc_call
    return frappe.handler.handle()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/handler.py", line 49, in handle
    data = execute_cmd(cmd)
           ^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/handler.py", line 85, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/__init__.py", line 1781, in call
    return fn(*args, **newargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/utils/typing_validations.py", line 31, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/www/printview.py", line 310, in get_html_and_style
    html = get_rendered_template(
           ^^^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/www/printview.py", line 224, in get_rendered_template
    html = frappe.get_attr(hook_func[-1])(jenv=jenv, template=template, print_format=print_format, args=args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "apps/frappe/frappe/utils/pdf.py", line 52, in pdf_body_html
    frappe.throw(
  File "apps/frappe/frappe/__init__.py", line 658, in throw
    msgprint(
  File "apps/frappe/frappe/__init__.py", line 623, in msgprint
    _raise_exception()
  File "apps/frappe/frappe/__init__.py", line 574, in _raise_exception
    raise exc
frappe.exceptions.PrintFormatError: Error in print format on line 1: cannot unpack non-iterable NoneType object

**Reply by mhaggag** [reply]
The formats need some work in error handling to present human readable error messages. What it's complaining about here is likely that there's no phase 1 business settings doctype for the invoice company. It's where it gets the information to put in the QR.

**Reply by ashish-greycube** [reply]
Yes, that makes sense

**Comment by ashish-greycube** [comment]
@mhaggag i did zatca phase 1 setup and SI shows in print format
![image](https://github.com/user-attachments/assets/6f18d905-6f90-45b1-83ba-0a018b72c538)

But when we do PDF, it gives error on cloud
related to missing fonts
![image](https://github.com/user-attachments/assets/2387a4e4-9a56-4aa5-a83f-f4d3dd428564)

>   File "apps/frappe/frappe/utils/print_format.py", line 234, in download_pdf
>     pdf_file = frappe.get_print(
>                ^^^^^^^^^^^^^^^^^
>   File "apps/frappe/frappe/__init__.py", line 2197, in get_print
>     return get_pdf(html, options=pdf_options, output=output) if as_pdf else html
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "apps/frappe/frappe/utils/pdf.py", line 98, in get_pdf
>     frappe.throw(_("PDF generation failed because of broken image links"))
>   File "apps/frappe/frappe/__init__.py", line 658, in throw
>     msgprint(
>   File "apps/frappe/frappe/__init__.py", line 623, in msgprint
>     _raise_exception()
>   File "apps/frappe/frappe/__init__.py", line 574, in _raise_exception
>     raise exc
> frappe.exceptions.ValidationError: PDF generation failed because of broken image links

**Reply by mhaggag** [reply]
We don't reference any fonts in the format. These look like warnings, not errors. Hard to say if they're fatal without seeing the full error output.

Can you check your letter head and confirm it doesn't reference non-existent or private images?
